### PR TITLE
Upgrade Kafka read/write transforms without upgrading the pipeline

### PR DESCRIPTION
**Pull request by chamikaramj**
This allows pipeline authors to upgrade Beam Kafka read/write transforms to a newer Beam version while keeping the rest of the pipeline in an older Beam version.

This will be useful for pipeline authors who wish to use newer releases of Kafka read/write transforms due to bug fixes, perf. improvements etc. but are unable to upgrade the full pipeline (for example, due to organizational restrictions).

Issue is https://github.com/apache/beam/issues/27943

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by chamikaramj** [comment]
assign set of reviewers

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 1]
```diff
@@ -328,4 +332,40 @@ private static int findAvailablePort() throws IOException {
...
+   *
+   * @param transform transform to lookup.
+   * @return a URN if discovered. Returns {@code null} otherwise.
+   */
+  @SuppressWarnings({
+    "rawtypes",
+    "EqualsIncompatibleType",
+  })
+  public static @Nullable String findUpgradeURN(
+      org.apache.beam.sdk.transforms.PTransform transform) {
```
Comment (johnjcasey) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:355]: can we fix this import?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java [hunk 2]
```diff
@@ -328,4 +332,40 @@ private static int findAvailablePort() throws IOException {
...
+    for (TransformPayloadTranslatorRegistrar registrar :
+        ServiceLoader.load(TransformPayloadTranslatorRegistrar.class)) {
+
+      for (Entry<
+              ? extends Class<? extends org.apache.beam.sdk.transforms.PTransform>,
+              ? extends TransformPayloadTranslator>
+          entry : registrar.getTransformPayloadTranslators().entrySet()) {
+        if (entry.getKey().equals(transform.getClass())) {
+          return entry.getValue().getUrn();
```
Comment (johnjcasey) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java:364]: Do we care if there are multiple potential matches?

File: sdks/java/io/kafka/upgrade/src/main/java/org/apache/beam/sdk/io/kafka/upgrade/KafkaIOTranslation.java [hunk 3]
```diff
@@ -0,0 +1,578 @@
+/*
```
Comment (johnjcasey) [inline sdks/java/io/kafka/upgrade/src/main/java/org/apache/beam/sdk/io/kafka/upgrade/KafkaIOTranslation.java:1]: This class looks like it contains a bunch of relatively boilerplate code. Given that we would want translation and upgrades for all pre-built composites eventually, is there any way to try and avoid this boilerplate?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java (hunk 1 above)
Comment (chamikaramj) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java]: Updated.

File: sdks/java/io/kafka/upgrade/src/main/java/org/apache/beam/sdk/io/kafka/upgrade/KafkaIOTranslation.java (hunk 3 above)
Comment (chamikaramj) [reply sdks/java/io/kafka/upgrade/src/main/java/org/apache/beam/sdk/io/kafka/upgrade/KafkaIOTranslation.java:1]: So all we do here is building a Row object from the transform object and vice versa using utilities available in the Beam Schema library. The rest of the boilerplate here is mostly error/null checks which cannot be refactored away easily. We do have fromByteArray/toByteArray methods here to concert objects that do not have a Schema defined (for example, user specified functions) but these are pretty small and different classes can choose to do the serialization in different ways. 

Do you have a specific refactoring in mind to simplify this code ?

**Review by chamikaramj — COMMENTED** [review]
Thanks.

**Comment by chamikaramj** [comment]
PTAL

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Review by johnjcasey — APPROVED** [review]

**Comment by chamikaramj** [comment]
Thanks.

**Pull request merged**